### PR TITLE
refactor(vector): rework feature support

### LINKED ISSUE #688: (big) multipolygon does not seem to be displayed correctly
A big multipolygon is not rendered correctly.

It is originally served from a  wfs  from geoserver, but I'm providing a geojson file of the multipolygon.
[multipolygon.zip](https://github.com/iTowns/itowns/files/1794114/multipolygon.zip)

It is correctly displayed in Qgis as shown in the screen captures below
**itowns**
multipolygon_itowns

**Qgis**
multipolygon_qgis

## Environment

* Version used: "itowns": "^2.3.0"
* Browser Name and version: firefox 58.0.2 /chromium 64.0
* Operating System and version (desktop or mobile): Linux Ubuntu 16.04

### PR DESCRIPTION
**Pull request by peppsac**
This commits refactors Feature related code.

It introduces a documented Feature object.
GeoJSON2Feature returns either a Feature or an array of Feature (when parsing
a FeatureCollection object).

Geometry merges now only happens for multi-* geometries to simplify user
handling of individual features.

Geometry holes are now supported:
capture d ecran de 2018-03-13 13-52-32

And last but not least this commit fixes the geometry merge bug from issue 688.

Fixes #688. This is a screenshot of the example geojson, displayed both as a texture (in yellow) and as a mesh (blue wireframe):

capture d ecran de 2018-03-13 13-54-23

BREAKING CHANGE: feature collection are not merged anymore
BREAKING CHANGE: GeoJSON2Feature now returns a Promise

### CODE REVIEW

**Comment by peppsac** [comment]
Notes:
  - this PR probably makes PR #673 useless.
  - this PR allows to color features individually (since a feature = a mesh)

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js:274]: why do you need a `Promise.resolve`?

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js]: nits `Coordinates`

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js]: `representing`

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js]: `all the geometries` or `FeatureGeometry`

Comment (autra) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Or all the `FeatureGeometry`

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js:268]: See below, I don't see why you need a promise here. Or you want to prepare the way for workers?

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js]: Why not returning `collec` everytime? At worst it'd be an empty array, maybe nicer to work with from the caller perspective...

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js]: (nits: I prefer the positive style, if possible)

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [inline src/Renderer/ThreeExtended/GeoJSON2Features.js]: copy-paste leftover :-)

Comment (autra) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Not sure it's still true for `multilineString` below anyway....

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: Better to use `Array.isArray` (see [this](http://web.mit.edu/jwalden/www/isArray.html) for more details). Here, this *should* work for our internal use case, but user can import Feature2Mesh directly so...
(also, babel/webpack heavily modifies `Array.prototype`, and I'm not sure this'll always work in all situations considering our stack or if we modify babel conf).

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: The jsdoc should be updated (apparently it hasn't been for a while).

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js:431]: could you add a doc on this function while you're at it?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js:365]: I'm wondering if it won't be simpler to have `geometryToMulti*` functions here. They don't share that much of common code.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: `Array.isArray`

Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: I don't get this. You can have several altitude per Coordinates? Why is the altitude changing the length of vertices? (I missed something somewhere)

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: There's no geometry.contour for lines right?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: `holes` -> `holesOffsets` ?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: `geometry[i].contour.offset` is always 0 with the current code right? Added here for the sake of consistency/robustness in case the contour is not at the beginning of the array?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Oh ok, it's because altitude can contains the extrude information. So actually `altitude.length` is < 2 in all cases. Maybe consider having a `extrude` param instead? If we hardcode the `extrude` logic, we could do it entirely. WDYT?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: If I understand this correctly, you have a slightly different color for the extruded face, is that it? If so why don't we rely on lights and materials for this?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: So we assume that other geometries will have the same orientation, right? Is it safe to assume according to you?

File: test/featureUtils_unit_test.js
Comment (autra) [inline test/featureUtils_unit_test.js]: With newer version of `mocha`, you can directly return a promise, this will have the same effect than the `done` mechanism.

Comment (autra) [reply test/featureUtils_unit_test.js]: But actually, GeoJSON2Features could return a plain object, see above.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (autra) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js:274]: (note: if it's because of the providers, they don't need it, they already wrap stuff in promises)

**Review by autra — CHANGES_REQUESTED** [review]
Nice!! I do have some nits and questions, but overall LGTM already. Thanks for this work :-)

**Comment by autra** [comment]
> this PR allows to color features individually (since a feature = a mesh)

I actually do have this need currently. Also, I feel that a feature = a mesh is more aligned with the spirit of the GeoJSON spec.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js:274]: No. It's simply to expose an async API.
(and then later, it opens the possibility to use WebWorker to parse the geojson, without changing again the API: see https://github.com/peppsac/itowns2/commit/729255b16cdf3d9e47745c3b750b997739c2ceeb#diff-df7974827a71f76f18b31a30dbc160b2R79 for an quick & dirty implementation)

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Fixed.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Fixed.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Fixed.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Ok, I modified the code to always return `collec`.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Fixed.

File: src/Renderer/ThreeExtended/GeoJSON2Features.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/GeoJSON2Features.js]: Fixed.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Thanks, I didn't know that `instanceof Array` had weird corner cases. Fixed.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Fixed.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Right, fixed. Thanks

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Yep

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: This comes from the original extrude code and I didn't want to change it.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Yeah that's the assumption and I feel it's safe (ie: I assume that a server will always return geometries with the same orientation).

File: test/featureUtils_unit_test.js
Comment (peppsac) [reply test/featureUtils_unit_test.js]: > With newer version of mocha

Does our version of `mocha` support this?

**Comment by peppsac** [comment]
Thanks for the review @autra 
I updated the PR and addressed all your comments.

File: test/featureUtils_unit_test.js
Comment (autra) [reply test/featureUtils_unit_test.js]: Yes, I did it for the `getCapabilities` PR

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (autra) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: Looks like a hack to me. I'd vote to remove this (but that can be in another PR).

**Comment by autra** [comment]
So I have some errors in `globe_wfs_extruded`:
```
TypeError: itowns.DEMUtils.getElevationValueAt(...) is undefined
Stack trace:
altitudePoint@http://localhost:8080/examples/globe_wfs_extruded.js:178:16
getAltitude@webpack-internal:///283:13:20
geometryToPoint@webpack-internal:///283:154:20
geometryToMesh@webpack-internal:///283:423:24
featureToThree@webpack-internal:///283:452:16
featureCollectionToThree@webpack-internal:///283:468:24
_convert@webpack-internal:///283:511:24
getFeatures/<@webpack-internal:///779:126:28
```
and
```
TypeError: result is undefined
Stack trace:
altitudeLine@http://localhost:8080/examples/globe_wfs_extruded.js:48:21
getAltitude@webpack-internal:///283:13:20
geometryToLine@webpack-internal:///283:166:20
geometryToMesh@webpack-internal:///283:429:24
featureToThree@webpack-internal:///283:452:16
featureCollectionToThree@webpack-internal:///283:468:24
_convert@webpack-internal:///283:511:24
getFeatures/<@webpack-internal:///779:126:28
```

**Comment by autra** [comment]
Forget this, seems like I didn't pull (maybe)

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (zarov) [inline src/Renderer/ThreeExtended/Feature2Mesh.js:357]: Not a big fan as `multiGeometries` is an array.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (zarov) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: If you try the `wfs` example, you'll get some error: `glDrawElements: attempt to access out of range vertices in attribute 0`. This is caused by the `indices` array being false, and from what I've seen it is only when `triangles` is empty at this stage.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (zarov) [inline src/Renderer/ThreeExtended/Feature2Mesh.js:163]: In `globe_wfs_extruded`, it seems that `altitude` resolves as an array, making everything a `NaN` afterwards.

**Review by zarov — CHANGES_REQUESTED** [review]
Thanks for doing this, I encountered some problems unfortunately.

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js:357]: I don't like this either but do you have any other proposal?

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js:163]: The `altitude` is an array case is supposedly handled here: https://github.com/iTowns/itowns/pull/691/files#diff-2a78502be7a11a150dcce98113215193R68

So I'm not sure what's going wrong.

Comment (zarov) [reply src/Renderer/ThreeExtended/Feature2Mesh.js:163]: My bad, I was working with a previous version. All good for this !

File: src/Renderer/ThreeExtended/Feature2Mesh.js
Comment (zarov) [reply src/Renderer/ThreeExtended/Feature2Mesh.js:357]: Either make it an object, containing the array and `vertices`, or make `vertices` available outside this condition to make it available farther below.

Comment (peppsac) [reply src/Renderer/ThreeExtended/Feature2Mesh.js:357]: Oh, right :)

I made it an object.

**Comment by peppsac** [comment]
I added a pair of unit test (with and without holes), validating the Mesh built are correct by computing the area.

**Review by autra — APPROVED** [review]

**Comment by autra** [comment]
LGTM. I pushed 2 commits, please r? (and squash) them before merging:
- one commit fixing the errors I've seen in console. It was probably already there before this PR btw. I *think* it happens when we load the geojson before an elevation is loaded there. There might be a better way of fixing this though.
- one commit fixing some nits (Array.isArray and inverting some loops to have positive logics).

Thanks!

**Review by zarov — APPROVED** [review]

**Pull request merged**
